README.md
           Comic Strip - Extract individual frames of a comic book
           -------------------------------------------------------


What is it?
-----------

While it is easy to read comic books on a regular PC, it is much more
inconvenient to do so on a PDA because of its small screen size - one typically
ends up scrolling around the whole page to read it. This application breaks down
each page of a comic book into individual frames (i.e. each frame becomes an
individual page) which can then be conveniently viewed on small-screen devices
like PDAs. It can process a single image file or a whole comic book file (only
cbz files are supported currently)

Installation
------------
```bash
git clone https://github.com/Denperidge/comicstrip.git
cd comicstrip
python3 -m pip install -r requirements.txt
python3 -m src.comicstrip
```

To run tests, simply use `python3 -m unittest`


Basic Terminology Used
----------------------

A typical page of a comic book consists of a number of frames separated by
horizontal/vertical white spaces called gutters. The title page might contain an
additional heading in the beginning. The height of each frame in a page (indeed
in all pages of a book) are more-or-less the same. The frame widths, however,
usually differ.

A comic book is usually a compressed archive of images with different
extensions to indicate the type of compression. Common formats include .cbz
(which is simply a zip archive of images) and .cbr (which is an RAR archive of
images). As mentioned above, comicstrip supports only .cbz files currently.


Usage
-----

Basic Invocation:

python3 -m src.comicstrip [options] [pgfile1, pgfile2, ...]

The comicstrip application is, at the moment, a command line application.
All parameters are passed in via command line parameters. A GUI front-end is
planned down the line to make the application easier to use.

Valid Options are:

--version
    Show program's version number and exit.

-h, --help
    Show a brief help message and exit.

-q, --quiet
    Don't print progress messages to stdout [default:False].
    By default a single dot is printed for every page processed. Every 5th page
    a page number is displayed (....5....10...etc)

-d, --debug           Enable debug prints [default:False].
    Used only for debugging. Ordinary users won't need to enable this. It would
    be helpful to enable '-q' when using the -d option so that the regular
    progress display does not interfere with the debug output (and vice versa).

-f FILE, --file=FILE
    Name of the input file. Mandatory parameter
    This could be the name of an image file or the name of a .cbz file.
    comicstrip supports the image formats supported by the Python Imaging
    Library.

--prefix=PREFIX
    Prefix for output files. Mandatory parameter
    The individual frames are saved to files with names of the form:
        `<prefix>00.jpg,`
        `<prefix>01.jpg`
        `<prefix>02.jpg`
        ...
    The number of leading zeroes is automatically adjusted depending on the
    number of output frames. If you want to save the output files into a
    particular directory just add the directory prefix to the file prefix (e.g.
    --prefix /tmp/foo-)

--left-ignore=PIXELS
    How much of the left margin to ignore when detecting rows [default:0]
    Sometimes (especially for scanned comics) the edges in the middle of a page
    tend to have a shadow all along it, making the gutter "non white". This
    interferes with the gutter detection algorithm and prevents a gutter from
    being successfully detected. This parameter tells the application how much
    many pixels on the left side should be ignored when detecting gutters. Note:
    This doesn't mean the pixels on the left of the margin are discarded in the
    final output - it just means that they are not considered during gutter
    detection.

--right-ignore=PIXELS
    How much of the right margin to ignore when detecting rows [default:0].
    Same as "--left-ignore" except that this is for the right side.

--firstpage=PGFILENAME
    Name of the title page in comic archive file.
    The first page of a comic archive could be different from other pages in
    that it might have a heading on top of the page (which needs to be skipped).
    This parameter tells the application the name of the first page if the input
    file is a comic book archive. Note: This parameter is ignored if the input
    file is a single image file and not a comic book archive.

--firstpg-row=PIXELS
    From which line of the first page should the processing start [default:0]
    This parameter tells the application whether or not there is a title in the
    first page and, if so, how many pixels to skip to ignore the title. Without
    this parameter the processing might stop at the title itself instead of
    skipping over it.

--startrow=PIXELS
    From which line of the each page (other than the first page) should the
    processing start [default:0]
    For scanned comics, the gutter on the top of the page might have a shadow
    along it, thus interfering with the gutter processing algorithm. This
    parameter tells the application how many pixels to skip when detecting rows
    in a page, thus allowing the algorithm to skip over the shadow.

--glob=GLOB
    A glob expression to select files to be processed from the book. (Not
    required if a file list is provided.)
    File names of pages of a comic book archive typically have a pattern.
    Instead of specifying each page by name, this parameter allows a glob
    pattern to be specified (e.g. --glob '*[0-9][0-9].jpg'). Remember to protect
    the glob expression in quotes to prevent file expansion in the shell.

--gutter-width=WIDTH
    Minimum width of the gutter [default:15]

--min-width=WIDTH
    Minimum width of a frame [default:100]
    A more accurate value speeds up the frame extraction algorithm

--min-height=HEIGHT
    Minimum height of a frame [default:100]
    A more accurate value speeds up the frame extraction algorithm

pgfile1, pgfile2, ... is a list of the names of the page files in the comic book
archive.  These can be left out if the "--glob" parameter is supplied. On the
other hand, if you wish to extract only a few pages of a comic book archive,
then they can be specified on the command line.

Of course, for single page image files, this list is empty.


Limitations
-----------
The application might not be able to successfully extract frames from a
(slightly) rotated page (i.e. the page is not entirely horizontal).

In some pages, some images "overflow" into the gutter space (e.g. some
speech bubbles overflow into the vertical gutter). In such cases the application
will not be able to successfully separate the two frames. If it is the vertical
gutter that is "occupied" then the two adjacent frames will be extracted as one.
If, however, the horizontal gutter is "occupied" then the two rows will be
treated as one!

If the gutters in a page are not "clean" (i.e. they contain some "random",
dark pixels - typical in some scanned images of old comics) then they too might
negatively impact the gutter processing algorithm. For images where the gutter
is "clean" (all "light" colors) there shouldn't be any problems.


Contact
-------
You can reach me at: koofoss(at)g-m-x(dot)com
(remove the '-'s between the characters above).


Happy Reading!
Koo.



changelog
Changelog:

2009/10/10  David Koo  <koofoss(at)g-m-x(dot)com>
	* Initial working copy. Works only with cbz files (no cbr support).

2011/09/11  Robin Monks  <devlinks(at)gmail(dot)com>
	* Clone from launchpad to GitHub

2018/02/22 	Soham Sil  <sil(dot)soham(at)gmail(dot)com>
	* Created Python3 compatible version

2023/08/05 	Denperidge  <denperidge(at)gmail(dot)com>
	* Restructured into library


test.py
import unittest

from src.tests.test_page import TestPage


if __name__ == "__main__":
    suites = []
    for suite in [
        TestPage
    ]:
        suites.append(unittest.makeSuite(suite))

    testSuite = unittest.TestSuite(suites)
    testSuite.run()

src/tests/helpers.py
from glob import glob
from os import makedirs
from os.path import dirname, join



test_data = join(dirname(__file__), "test_data/")

test_jpg = join(test_data, "test.jpg")

test_originals_dir = join(test_data, "output_from_original/")
test_output_dir = join(test_data, "output_from_tests/")

makedirs(test_output_dir, exist_ok=True)



def list_original_images():
    return glob("*.jpg", root_dir=test_originals_dir)

def list_generated_images():
    return glob("*.jpg", root_dir=test_output_dir)

def read_image(path):
    with open(path, "rb") as file:
        content = file.read()
    
    return content

def read_images(path1, path2):
    return [read_image(path1), read_image(path2)]

def read_test_images(filename):
    return read_images(
        test_originals_dir + filename,
        test_output_dir + filename
    )


src/tests/test_page.py
import unittest

from ..comicstrip.Page import Page
from .helpers import test_jpg, test_data, read_test_images, list_original_images, list_generated_images

class TestPage(unittest.TestCase):
    
    def test_save(self):
        Page(test_jpg).save(test_data + "output_from_tests/cstrip-")

        original_images = list_original_images()
        generated_images = list_generated_images()

        self.assertEqual(
            len(original_images), len(generated_images),
            f"There is a different amount of images generated ({len(generated_images)}) than originally ({len(original_images)})"
            )

        for filename in original_images:
            original_image, generated_image = read_test_images(filename)
            self.assertEqual(original_image, generated_image,
                             f"The generated output for {filename} is different than originally")
            

   


if __name__ == "__main__":
    unittest.main()

src/comicstrip/Page.py
from .utils import debug, nopfn
import sys
import inspect
import zipfile
from os.path import exists
import fnmatch
from math import log


from PIL import Image
from PIL.ImageFile import Parser
from PIL.ImageEnhance import Contrast
from PIL.ImageOps import autocontrast


# gutter color
gcolor = 255

# gutter width
gwidth, gheight = 10, 10

# we first change the contrast of the image (to remove noise in the gutters) and then digitize
# them. This tells us how much to set the contrast to
contrast = 0.8

# barrier (in terms of a scale from 1 to 255). Used in dividing the image into black and
# white portions - if a color is < barrier then it is converted to black else white. The
# numbers here come from trial and error!
barrier = 210


class Page():
    """A page of the book"""

    # Typical Layout of a page:
    #   +-Page-----------------------------------------+
    #   |                                              |<- Gutter
    #   | +----------+ +--------------+ +-----------+  |-----
    #   | |          | |              | |           |  | Row
    #   | |          | |              | |           |  |
    #   | +----------+ +--------------+ +-----------+  |-----
    #   |                                              |<- Gutter
    #   | +----------+ +--------------+ +-----------+  |-----
    #   | |          | |              | |           |  | Row
    #   | |          | |              | |           |  |
    #   | +----------+ +--------------+ +-----------+  |-----
    #   |                                              |<- Gutter
    #   +----------------------------------------------+
    #                <->
    #              Gutter
    #
    # The left right, top and bottom gutters of the page may not be present
    #
    # Algo:
    #   - Starting with a 'startRow' we keep moving down as long as the whole row is still a
    #     gutter. This gives us a row start
    #   - Now move down "fheight" pixels (where fheight is the minimum height of a frame)
    #     and keep moving down until a gutter is encountered (or bottom of page). That is the
    #     bottom of our row
    #
    # Once we have our rows we repeat a similar procedure for each row to get the frames
    # of each row. We could transpose the whole image, copy out each row and use the row
    # detection functions for extracting the columns, but we want to use as little copies as
    # possible.
    #
    # To make the detection of boundaries easier, we first increase the contrast and remove any
    # gradients. This, unfortunately involves making 2 copies of the image (PIL, unfortunately,
    # (for most part) does not support in-place operations) :(.
    
    def __init__(self, infile:str, startRow=0, lignore=0, rignore=0, contents=True, pgNum=1, quiet=False, debug=False, fwidth:int=50, fheight=50):
        """A page object.
        Valid keyword parameters:
        infile:
            A String holding page contents or the page file name (depending on the contents
            parameter)
        startRow (default:0):
            Which row to start analyzing from. This is typically used for analyzing the first
            page of a comic where some top part of the page contains the title (which we need
            to skip)
        lignore, rignore (default:0 for both):
            Scanned pages might have some non-white color on one or both of the edges which
            interfere with gutter detection. These paramters tell the gutter detection
            algorithm to adjust the left boundary by lignore and right boundary by
            rignore when locating gutters
        contents (default: True):
            True => "infile" is a string consisting of page contents
            False => "infile" is a string holding the name of the page file to open
        pgNum (default:1):
            Page number (used when processing a whole book)
        quiet:
            Don't print any status messages
        debug:
            Enable debug prints
        fwidth, fheight:
            Minimum width (height resp) of a frame"""
    
        self.infile = infile

        self.startRow = startRow

        self.lignore = lignore
        self.rignore = rignore

        self.contents = False if exists(self.infile) else True
        self.pgNum = pgNum
        self.quiet = quiet
        self.debug = debug
        
        self.fwidth = fwidth
        self.fheight = fheight
        
        quietFns = {False:(self._prnfn, self._nlfn), True:(nopfn, nopfn)}
        self.prnfn, self.nlfn = quietFns[self.quiet]
        if self.contents:
            parser = Parser()
            parser.feed(self.infile)
            self.orig = parser.close()
        else:
            self.orig = Image.open(self.infile)
        self.img = self._prepare()
        self.frames = self._getFrames()

    def _isGutterRow(self, left, row, right):
        """Is the row from [left, right) a gutter?"""
        nongutter = [x for x in range(left, right) if gcolor != self.img.getpixel((x,row))]
        return len(nongutter) == 0

    def _isGutterCol(self, col, top, bot):
        """Is the column from [top, bot) a gutter?"""
        nongutter = [r for r in range(top, bot) if gcolor != self.img.getpixel((col,r))]
        return len(nongutter) == 0

    def _getRow(self, l, startRow, r, b):
        """Get the first row of (l, startRow, r, b).
        We don't use page members directly 'coz this function is called to further refine each
        frame's top and bottom boundaries"""
        debug(self.debug, "startRow:", startRow)
        if startRow >= b:
            return (-1,-1)

        # move down as long as the entire row (within window) is "gutter color"
        row1 = startRow
        while row1 < b and self._isGutterRow(l, row1, r):
            row1 += 1
        debug(self.debug, "row1:", row1)
        if row1 == b:
            return (-1, -1) # We've finished the image, no more rows

        # Now to find the bottom gutter - we assume a frame at least fheight pixels
        # high so we'll just skip those many pixels
        row2 = row1 + self.fheight
        debug(self.debug, "row2 starting with:", row2)
        if row2 > b:
            return (-1, -1) # probably looking at the area after the last row (e.g. pagenum)
        while row2 < b and not self._isGutterRow(l, row2, r):
            row2 += 1

        debug(self.debug, "row2:", row2)
        if row2 - row1 < self.fheight:
            return (-1, -1) # not a proper frame (e.g. contains pagenum)
        return (row1, row2)

    def _prnfn(self, symbol):
        print(symbol),
        sys.stdout.flush()

    def _nlfn(self):
        print

    def _getRows(self, startRow):
        """Get a list of all rows starting from startRow. Display progress if indicated"""
        top, rows = startRow, []
        count = 0
        l,r,b = self.lignore, self.img.size[0] - self.rignore, self.img.size[1] - 1
        while True:
            top, bot = self._getRow(l, top, r, b)
            if top != -1:
                debug(self.debug, "got row:", top, bot)
                rows.append((0, top, self.img.size[0]-1, bot))
                top = bot + (gheight//2)
                count += 1
            else:
                debug(self.debug, "No more rows")
                break
        debug(self.debug, "rows:", rows)
        return rows

    def _getCol(self, startCol, t, b):
        """Get leftmost column of a row starting from startCol.
        The row is enclosed by t and b."""
        debug(self.debug, "startCol, t, b:", startCol, t, b)
        r = self.img.size[0] - 1
        if startCol >= r:
            return (-1,-1)

        # move right as long as the entire column (within window) is "gutter color"
        col1 = startCol
        while col1 < r and self._isGutterCol(col1, t, b):
            col1 += 1
        if col1 == r:
            return (-1, -1) # We've finished the row, no more columns
        debug(self.debug, "col1:", col1)

        # Now to find the right boundary of the frame
        col2 = col1 + self.fwidth
        debug(self.debug, "Starting with column:", col2)
        if col2 > r:
            return (-1, -1) # no frame here - just gutter area on the right
        while col2 < r and not self._isGutterCol(col2, t, b):
            col2 += 1
        debug(self.debug, "col2:", col2)

        if col2 - col1 < self.fwidth:
            return (-1, -1) # not a proper frame
        return (col1, col2)

    def _getCols(self, rt, rb):
        """Get columns from row bounded on top and bottom by rt & rb."""
        left, cols = 0, []
        while True:
            left, right = self._getCol(left, rt, rb)
            if left != -1:
                debug(self.debug, "got column:", left, right)
                cols.append((left, rt, right, rb))
                left = right + (gwidth//2)
            else:
                debug(self.debug, "No more columns")
                break
        debug(self.debug, "cols:", cols)
        return cols

    def _getFrames(self):
        """Get all frames in page"""
        # Display progress in the form of ....Pg....Pg.... (1 dot per page)
        if self.pgNum % 5 == 0:
            symbol = str(self.pgNum)
        else:
            symbol = "."
        self.prnfn(symbol)
        # first get all the rows, traversing the entire height of the image (after
        # accounting for the adjustments
        rows = self._getRows(self.startRow)
        debug(self.debug, "Got rows:", rows)

        # now get columns in each row
        frames = []
        for rl, rt, rr, rb in rows:
            debug(self.debug, "Row:", rl, rt, rr, rb)
            cols = self._getCols(rt, rb)
            debug(self.debug, "Got Columns:", cols)
            frames.extend(cols)

        debug(self.debug, "=== Frames:", frames)
        # Now try to further trim the top and bottom gutters of each frame (left and right
        # gutters would already be as tight as possible) and then extract the area from the
        # original image
        fimgs = []
        for (fl, ft, fr, fb) in frames:
            debug(self.debug, "Refining:", fl, ft, fr, fb)
            newt, newb = self._getRow(fl, ft, fr, fb)
            if newt == -1:
                # The frame is already as tight as possible
                debug(self.debug, "Cannot reduce any further")
                newt, newb = ft, fb
            else:
                debug(self.debug, "Got:", newt, newb)
            fimg = Image.new("RGB", (fr - fl, newb - newt))
            fimg = Image.new("RGB", (fr - fl, newb - newt))
            fimg.paste(self.orig.crop((fl, newt, fr, newb)), (0, 0))
            fimgs.append(fimg)
        return fimgs

    def _digitize(self, color):
        if color // barrier == 0:
            result = 0
        else:
            result = 255
        return result

    def _prepare(self):
        bwimg = self.orig.convert("L")
        return Contrast(autocontrast(bwimg, 10)).enhance(contrast).point(self._digitize)

    def save(self, prefix="", counter=0):
        debug(self.debug, "Saving pages:")
        npfxDigits = int(log(len(self.frames), 10)) + 1
        for fimg in self.frames:
            fname = "%s%0*d.jpg" % (prefix, npfxDigits, counter)
            debug(self.debug, "    Saving:", fname)
            fimg.save(fname)
            counter += 1
        return counter
    

def pgtest():
    global DEBUG
    # DEBUG = True
    if len(sys.argv) == 3:
        startrow = int(sys.argv[1])
    else:
        startrow = 0
    pgs = sys.argv[2:]
    counter = 11
    for pgname in pgs:
        print("Processing page:", pgname)
        pg = Page(pgname, startrow, 50, 50)
        counter = pg.save("ottokar-", counter)

src/comicstrip/Comic.py
from .Page import Page
import zipfile
import fnmatch


class Comic(object):
    """A comic book - a cbz or a cbr file"""

    PROCESS_BOOK = 0
    PROCESS_PAGE = 1

    keys = ["infile", "prefix", "firstPg", "firstPgRow", "startRow", "lignore", "rignore",
            "filePat", "fileList", "quiet", "gwidth", "debug", "fwidth", "fheight"]
    pgkeys = ["startRow", "lignore", "rignore", "infile", "quiet", "debug", "fwidth", "fheight"]

    def __init__(self, **kw):
        """Initialize a comic book (cbz) file.
        Permitted arguments
        infile:
            Name of the comic book to process. Only cbz files are supported at the moment.
        prefix:
            Output file prefix. Each output fill will be of the form prefixNNN
        firstPg:
            The name of the first page to start analyzing from.
        firstPgRow:
            The first page might start with a title first. We need to skip over this to
            get to the actual rows
        lignore, rignore:
            Sometimes scanned pages are not at the edges. lignore, rignore tell us
            how many left/right columns to ignore when searching for rows
        filePat:
            File names in the archive that match this pattern will be processed. The pattern is
            a glob expression. If both fileList and filePat are specified, fileList is used, if
            neither are specified then the file pattern "*.jpg" is used.
        fileList:
            File names in the archive that are in this list will be processed. If both fileList
            and filePat are specified, fileList is used, if neither are specified then the file
            pattern "*.jpg" is used.
        gwidth:
            Minimum width (and height) of the gutter.
        fwidth, fheight:
            Minimum width (height resp) of a frame.
        quiet:
            Don't print any progress messages.
        debug:
            Enable debug prints"""
        object.__init__(self)
        [self.__setattr__(k, kw[k]) for k in Comic.keys]
        self.counter = 0 # used when writing output
        try:
            self.zfile = zipfile.ZipFile(kw["infile"])
        except:
            # is probably a single page image instead of a comic book
            self.actionType = Comic.PROCESS_PAGE
        else:
            self.actionType = Comic.PROCESS_BOOK

    def processBook(self):
        if len(self.fileList) == 0:
            self.fileList = fnmatch.filter(self.zfile.namelist(), self.filePat)
            self.fileList.remove(self.firstPg)
        kw = dict([(k, object.__getattribute__(self, k)) for k in Comic.pgkeys])
        kw["pgNum"] = 1
        kw["contents"] = True
        if self.firstPg:
            # process the start page separately
            buf = self.zfile.read(self.firstPg)
            kw["startRow"] = self.firstPgRow
            kw["infile"] = buf
            pg = Page(**kw)
            self.counter = pg.save(self.prefix, self.counter)
            kw["pgNum"] += 1

        # for other pages, startRow = startRow
        kw["startRow"] = self.startRow
        for fname in self.fileList:
            buf = self.zfile.read(fname)
            kw["infile"] = buf
            pg = Page(**kw)
            self.counter = pg.save(self.prefix, self.counter)
            kw["pgNum"] += 1

    def processPg(self):
        kw = dict([(k, object.__getattribute__(self, k)) for k in Comic.pgkeys])
        kw["pgNum"] = 1
        kw["contents"] = False
        Page(**kw).save(self.prefix, self.counter)

    def process(self):
        if self.actionType == Comic.PROCESS_BOOK:
            self.processBook()
        else:
            self.processPg()
        print


src/comicstrip/cli.py

from .Comic import Comic
from argparse import ArgumentParser, FileType
from .constants import _VERSION

def getargs(parser):
    (options, args) = parser.parse_args()
    kw = {}
    kw["infile"] = options.infile
    if kw["infile"] is None:
     raise ValueError("Input File Not Specified")
    kw["prefix"] = options.prefix
    kw["firstPg"] = options.firstPg
    kw["firstPgRow"] = options.firstPgRow
    kw["startRow"] = options.startRow
    kw["lignore"] = options.lignore
    kw["rignore"] = options.rignore
    kw["filePat"] = options.filePat
    kw["quiet"] = options.quiet
    kw["gwidth"] = options.gwidth
    kw["fwidth"] = options.fwidth
    kw["fheight"] = options.fheight
    kw["debug"] = options.debug
    kw["fileList"] = args
    return kw


parser = ArgumentParser(usage="%(prog)s [options] [pgfile1, pgfile2, ...]",
                        prog="comicstrip",
                        description="Split a comic page into individual frames")

parser.add_argument("-v", "--version",
                    action="version",
                    version="%%(prog)s %s" % (_VERSION))

parser.add_argument("-q", "--quiet", 
                  action="store_true", 
                  dest="quiet",
                  help="Don't print progress messages to stdout [default:%default]")

parser.add_argument("-d", "--debug", 
                  dest="debug", 
                  action="store_true",
                  help="Enable debug prints [default:%default]")

parser.add_argument("-f", "--file", 
                    dest="infile",
                    type=FileType("rb"),
                    metavar="FILE",
                    help="Name of the input file",
                    required=True)

parser.add_argument("--prefix", 
                    dest="prefix",
                    help="Prefix for output files")

parser.add_argument("--left-ignore", 
                    type=int,
                    dest="lignore",
                    metavar="PIXELS",
                    help="How much of the left margin to ignore when detecting rows [default:%default]")

parser.add_argument("--right-ignore", 
                    type=int, 
                    dest="rignore", 
                    metavar="PIXELS",
                    help="How much of the right margin to ignore when detecting rows [default:%default]")

parser.add_argument("--firstpage", 
                    dest="firstPg", 
                    type=str, 
                    metavar="PGFILENAME",
                    help="Name of the title page in comic archive file")

parser.add_argument("--firstpg-row", 
                    type=int, 
                    dest="firstPgRow", 
                    metavar="PIXELS",
                    help="From which line of the first page should the processing start [default:%default]")

parser.add_argument("--startrow", 
                    type=int, 
                    dest="startRow", 
                    metavar="PIXELS",
                    help="From which line of the each page (other than the first page) should the processing start [default:%default]")

parser.add_argument("--glob", 
                    dest="filePat", 
                    metavar="GLOB",
                    help="A glob expression to select files to be processed from the book. (Not required if a file list is provided.)")

parser.add_argument("--gutter-width", 
                    dest="gwidth", 
                    metavar="WIDTH",
                    help="Minimum width of the gutter [default:%default]")

parser.add_argument("--min-width",
                    dest="fwidth", 
                    metavar="WIDTH", 
                    type=int,
                    help="Minimum width of a frame [default:%default]")

parser.add_argument("--min-height",
                    dest="fheight",
                    metavar="HEIGHT",
                    type=int,
                    help="Minimum height of a frame [default:%default]")

parser.add_argument("fileList",
                    nargs="*",
                    type=FileType("rb"),
                    help="")

parser.set_defaults(quiet=False,
                    prefix="cstrip-",
                    lignore=0,
                    rignore=0,
                    firstPgRow=0,
                    startRow=0,
                    gwidth=15,
                    fwidth=50,
                    fheight=50,
                    debug=True)


def run_cli():
    kw = vars(parser.parse_args())
    book = Comic(**kw)
    book.process()
